### PR TITLE
[3.4.1] Update test certificates in 3.4.0

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Prepare for a patch release with updated certificates, plus one minor bugfix that was accidentally merged to master instead of development. Bump the library version number to 3.4.1.

Contents:

* Copy all test certificate data from `development`. (This includes files that are unused.)
* Make the minimum changes to X.509 unit tests so that they pass: some files have moved to the `parse_input` subdirectory; update a dependency on DES which is now AES.
* Bump the version to 3.4.1. Include a changelog entry.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** no: for 2.28 we'll directly make a release in `mbedtls-2.28`
- [ ] **tests** [`faketime 2024-01-16 all.sh`](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/36/) on a824f8bc913bcacf8358380b4f35769c7484935 →

### CODE REVIEW

**Review by yuhaoth — APPROVED** [review]
LGTM. I think the CI failure can be ignored

There are some failure in CI. I do not found unknown failure. I checked some files, they reports ASAN fail, that is known issue. And some logs report "address assign" issue, that might be environment issue.

About doc build failure, I am not sure the reason, please double check if it can be ignore.

File: tests/suites/test_suite_x509write.data
Comment (lpy4105) [inline tests/suites/test_suite_x509write.data:1]: There are three cases in `tests/suites/test_suite_pkparse.data` that need similar change , according to https://github.com/Mbed-TLS/mbedtls/commit/1a4cc5e92ca8dbffa5606e4da84eb02e1c70e83d.
Though, the three cases are not run by any test components on CI.

https://github.com/Mbed-TLS/mbedtls/blob/a824f8bc913bcacf8358380b4f35769c74849357/tests/suites/test_suite_pkparse.data#L1-L11

**Review by lpy4105 — COMMENTED** [review]
We might miss some changes from `development`, please check.

For the faketime tests, there are a lot of build failures like:
```
[2023-08-02T21:11:22.618Z]   CC    src/psa_exercise_key.c
[2023-08-02T21:11:22.618Z]   CC    pkey/pk_decrypt.c
[2023-08-02T21:11:22.874Z] ../tests/src/psa_exercise_key.o: file not recognized: File truncated
[2023-08-02T21:11:22.874Z] collect2: error: ld returned 1 exit status
[2023-08-02T21:11:22.874Z] Makefile:256: recipe for target 'pkey/pk_decrypt' failed
[2023-08-02T21:11:22.874Z] make[1]: *** [pkey/pk_decrypt] Error 1
[2023-08-02T21:11:22.874Z] Makefile:15: recipe for target 'programs' failed
[2023-08-02T21:11:22.874Z] make: *** [programs] Error 2
[2023-08-02T21:11:22.874Z] make: *** Waiting for unfinished jobs....
```
I think they might be caused by some issues of the old version of `faketime`. I tried some of the tests locally with self-built `faketime (v0.9.10)` and no similar issue has been found. So, many test components were not run with faketime. I'm not sure if we could ignore this.

File: tests/data_files/Makefile [hunk 1]
```diff
@@ -1452,6 +1820,72 @@ pkcs7_data_cert_signeddata_sha256.der: pkcs7_data_cert_signed_sha256.der
...
+					server2.key.der \
+					cli2.crt \
+					cli2.crt.der \
+					cli2.key \
+					cli2.key.der \
+					cli-rsa-sha256.crt \
+					cli-rsa-sha256.crt.der \
+					cli-rsa.key \
+					cli-rsa.key.der
+../src/test_certs.h: ../scripts/generate_test_cert_macros.py \
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/data_files/Makefile:1855]: Out of scope of the release. I just copied `test_certs.h` from `development`, I don't care about generating it in 3.4.1.

File: tests/suites/test_suite_x509write.data
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509write.data:1]: > Though, the three cases are not run by any test components on CI.

Thanks for noticing that. It's a bug, still a bug in development (I'll make a patch).

File: tests/suites/test_suite_pkparse.data [hunk 2]
```diff
@@ -987,7 +987,7 @@ depends_on:MBEDTLS_PEM_PARSE_C:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP192R1_ENABLED
 pk_parse_keyfile_ec:"data_files/ec_prv.sec1.comp.pem":"NULL":0
 
 Parse EC Key #3 (SEC1 PEM encrypted)
-depends_on:MBEDTLS_DES_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP192R1_ENABLED:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_HAS_MD5_VIA_LOWLEVEL_OR_PSA
+depends_on:MBEDTLS_DES_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP192R1_ENABLED:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_HAS_ALG_MD5_VIA_LOWLEVEL_OR_PSA
```
Comment (lpy4105) [inline tests/suites/test_suite_pkparse.data:990]: CI failed. Here the depenency on `MBEDTLS_HAS_ALG_MD5_VIA_LOWLEVEL_OR_PSA` is inconsistent with `PEM_RFC1421`:
https://github.com/Mbed-TLS/mbedtls/blob/7fdf5497ee05c87219c589edaa0d5de74e800cbb/library/pem.c#L51-L57

File: tests/suites/test_suite_x509write.data
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_x509write.data:1]: Fixing the misspelled dependency causes `test_psa_crypto_config_accel_hash` to fail. This is also the case in `development` (first commit of https://github.com/Mbed-TLS/mbedtls/pull/8029). So in 3.4.1 I'm going to make only the DES→AES change and not fix the dependency.

File: tests/suites/test_suite_pkparse.data (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pkparse.data]: [This is out of scope of 3.4.1](https://github.com/Mbed-TLS/mbedtls/pull/8022#discussion_r1282936509). I've amended the last commit to only do the DES→AES change and not fix an unrelated bug. Fixing that bug will happen in https://github.com/Mbed-TLS/mbedtls/pull/8029.

**Comment by daverodgman** [comment]
In order to give as much time as possible to finish all the release prep, I've prepared 3.4.1rc based on the current shape of this PR and am running the CI here: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-new/811/ - if further changes are needed, I'll re-do this.

**Comment by gilles-peskine-arm** [comment]
I don't anticipate any more changes unless something comes up in review.
a824f8bc913bcacf8358380b4f35769c74849357 passed the CI except for readthedocs which was handled by #8026. The subsequent commit 6919546ddfece3cc53f3e27382ee3976a4983cf9 should only touch test cases that aren't executed (due to a bug that's too hard to fix for an emergency release) so that should be fine.

**Comment by gilles-peskine-arm** [comment]
I haven't managed to get a meaningful CI run with faketime (https://github.com/Mbed-TLS/mbedtls/pull/7494 shows a way but it's a bit too much work for this emergency release). I've checked locally that `make test && tests/ssl-opt.sh && tests/compat.sh -V YES -m tls12 -t 'ECDSA RSA' -e 'CAMELLIA\|ARIA\|NULL'` pass with `faketime 2024-01-17` in the `full` configuration.

**Review by lpy4105 — APPROVED** [review]
LGTM

**Review by yanesca — APPROVED** [review]
Did an incremental review on the top of @yuhaoth 's and it looks good to me.

**Comment by daverodgman** [comment]
Confirmed that with this merged, `mbedtls-3.4.1-proposed` is identical to `restricted/mbedtls-3.4.1rc`

**Pull request merged**